tests/contrib/test_kubernetes/test_base.py

```diff
@@ -0,0 +1,124 @@
+    K8sEnv,
+)
+from mlem.core.objects import DeployStatus
+from tests.contrib.test_kubernetes.conftest import k8s_test
+from tests.contrib.test_kubernetes.utils import Command
+
+
+@pytest.fixture
+def minikube_env_variables():
+    output = subprocess.check_output(
+        ["minikube", "-p", "minikube", "docker-env"]
+    )
+    export_re = re.compile('export ([A-Z_]+)="(.*)"\\n')
+    export_pairs = export_re.findall(output.decode("UTF-8"))
+    for k, v in export_pairs:
+        os.environ[k] = v
+
+
+@pytest.fixture
+def load_kube_config():
+    config.load_kube_config(os.getenv("KUBECONFIG", default="~/.kube/config"))
+
+
+@pytest.fixture
+def k8s_deployment(minikube_env_variables, model_meta_saved_single):
+    return K8sDeployment(
+        name="ml",
+        model=model_meta_saved_single.make_link(),
+        image_pull_policy="Never",
+        service_type="LoadBalancer",
+        daemon=DockerDaemon(host=os.getenv("DOCKER_HOST", default="")),
+    )
+
+
+@pytest.fixture
+def docker_image(k8s_deployment):
+    tmpdir = tempfile.mkdtemp()
+    k8s_deployment.dump(os.path.join(tmpdir, "deploy"))
+    return build_model_image(
+        k8s_deployment.get_model(),
+        k8s_deployment.image_name,
+        k8s_deployment.server
+        or project_config(
+            k8s_deployment.loc.project if k8s_deployment.is_saved else None
+        ).server,
+        DockerEnv(
+            daemon=DockerDaemon(host=os.getenv("DOCKER_HOST", default=""))
+        ),
+        force_overwrite=True,
+    )
+
+
+@pytest.fixture
+def k8s_deployment_state(docker_image, model_meta_saved_single):
+    return K8sDeploymentState(
+        image=docker_image,
+        model_hash=model_meta_saved_single.meta_hash(),
+    )
+
+
+@pytest.fixture
+def k8s_env():
+    return K8sEnv()
+
+
+@k8s_test
+def test_deploy(
+    load_kube_config,  # pylint: disable=unused-argument
+    k8s_deployment,
+    k8s_deployment_state,
+    k8s_env,
+):
+    k8s_deployment.update_state(k8s_deployment_state)
+    assert k8s_env.get_status(k8s_deployment) == DeployStatus.NOT_DEPLOYED
+    k8s_env.deploy(k8s_deployment)
+    k8s_deployment.wait_for_status(
+        DeployStatus.RUNNING,
+        allowed_intermediate=[DeployStatus.STARTING],
+        timeout=10,
+        times=5,
+    )
+    time.sleep(5)
+    assert k8s_env.get_status(k8s_deployment) == DeployStatus.RUNNING
+    k8s_env.remove(k8s_deployment)
+    time.sleep(5)
+    assert k8s_env.get_status(k8s_deployment) == DeployStatus.NOT_DEPLOYED
+    with k8s_deployment.daemon.client() as client:
+        k8s_deployment_state.image.delete(client, force=True)
+    time.sleep(5)
+
+
+@k8s_test
+def test_deployed_service(
+    load_kube_config,  # pylint: disable=unused-argument
+    k8s_deployment,
+    k8s_deployment_state,
+    k8s_env,
+):
+    time.sleep(15)
+    k8s_deployment.update_state(k8s_deployment_state)
+    k8s_env.deploy(k8s_deployment)
+    cmd = Command("minikube tunnel")
+    cmd.run(timeout=20, shell=True)
```

Comment:
shouldnt this be stopped after or smth like that?

Reply:
yes it's stopped due to `timeout` as `20`
you are talking about `minikube tunnel` right?

Reply:
so it will stop after 20 seconds, the tunneling stuff basically

Reply:
so the test will always run for 20 seconds?

Reply:
I think test should stop immediately once it passes. The 20 seconds stuff is for minikube tunnel to stop, we don't want to be waiting for prediction forever, so it stops the tunneling in the background after 20 seconds which is enough for test to pass and stop.